### PR TITLE
Add popup strategy implementation, (sign in, link, reauth)

### PR DESCRIPTION
**Pull request by samhorlbeck**

### CODE REVIEW

File: packages-exp/auth-exp/demo/src/index.js [hunk 1]
```diff
@@ -1173,10 +1180,10 @@ function onPopupRedirectAddCustomParam(event) {
  * Performs the corresponding popup/redirect action for a generic provider.
  */
 function onPopupRedirectGenericProviderClick() {
-  alertNotImplemented();
-  // var providerId = $('#popup-redirect-generic-providerid').val();
-  // var provider = new firebase.auth.OAuthProvider(providerId);
-  // signInWithPopupRedirect(provider);
+  // alertNotImplemented();
```
Comment (avolkovi) [inline packages-exp/auth-exp/demo/src/index.js:1183]: don't need this line at all

File: packages-exp/auth-exp/src/core/strategies/popup.ts [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+interface PendingPromise {
+  resolve: (cred: UserCredential) => void;
+  reject: (error: Error) => void;
+}
+
+export async function signInWithPopup(
+  authExtern: externs.Auth,
+  provider: externs.AuthProvider,
+  resolverExtern: externs.PopupRedirectResolver
+): Promise<UserCredential> {
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/strategies/popup.ts:49]: externs.UserCredential

File: packages-exp/auth-exp/src/core/strategies/popup.ts [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+
+  await auth.updateCurrentUser(cred.user);
+  return cred;
+}
+
+export async function reauthenticateWithPopup(
+  userExtern: externs.User,
+  provider: externs.AuthProvider,
+  resolverExtern: externs.PopupRedirectResolver
+): Promise<UserCredential> {
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/strategies/popup.ts:70]: externs.UserCredential

File: packages-exp/auth-exp/src/core/strategies/popup.ts [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    user
+  );
+  return action.execute();
+}
+
+export async function linkWithPopup(
+  userExtern: externs.User,
+  provider: externs.AuthProvider,
+  resolverExtern: externs.PopupRedirectResolver
+): Promise<UserCredential> {
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/strategies/popup.ts:89]: externs.UserCredential

File: packages-exp/auth-exp/src/core/strategies/popup.ts [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+  const auth = authExtern as Auth;
+  const resolver = resolverExtern as PopupRedirectResolver;
+
+  const action = new PopupAction(
+    auth,
+    AuthEventType.SIGN_IN_VIA_POPUP,
+    _signIn,
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/strategies/popup.ts:56]: you don't need to pass the resolver, you can map the AuthEventType to the correct resolver and simplify your constructor

File: packages-exp/auth-exp/src/core/strategies/popup.ts [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+    provider,
+    resolver
+  );
+  const cred = await action.execute();
+
+  await auth.updateCurrentUser(cred.user);
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/strategies/popup.ts:62]: If you wait for my refactor, _signIn calls signInWithCredential and does the updateCurrentUser for you, then all three of these look the same

File: packages-exp/auth-exp/demo/src/index.js (hunk 1 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/demo/src/index.js]: Done

File: packages-exp/auth-exp/src/core/strategies/popup.ts (hunk 2 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/core/strategies/popup.ts]: Done

File: packages-exp/auth-exp/src/core/strategies/popup.ts (hunk 3 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/core/strategies/popup.ts]: Done

File: packages-exp/auth-exp/src/core/strategies/popup.ts (hunk 4 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/core/strategies/popup.ts]: Done

File: packages-exp/auth-exp/src/core/strategies/popup.ts (hunk 5 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/core/strategies/popup.ts]: Done

File: packages-exp/auth-exp/src/core/strategies/popup.ts (hunk 6 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/core/strategies/popup.ts]: Done

**Pull request merged**